### PR TITLE
[cxxmodules] Print stacktrace before aborting on a missing exception.

### PR DESCRIPTION
**Pull request by Teemperor**
We will probably see an increasing amount of these failures with
C++ modules as we now deserialize all declarations instead of just
the PCH ones. To safe us a lot of debugging time on where to push
the needed transaction, let's directly print the stack trace here
in the rare case that we crash here.

### CODE REVIEW

**Comment by Teemperor** [comment]
@phsft-bot build

**Comment by vgvassilev** [comment]
@phsft-bot build with flags -Dclingtest=On

**Comment by Teemperor** [comment]
The failing count include test is also failing in master, so this is not a regression: https://github.com/root-project/root/pull/962#issuecomment-328043968

**Comment by Teemperor** [comment]
@phsft-bot build with flags -Dclingtest=Off

let's make sure it's not breaking master

File: interpreter/cling/lib/Interpreter/DeclCollector.cpp [hunk 1]
```diff
@@ -57,6 +58,15 @@ namespace {
         return true;
     return false;
   }
+
+  /// \brief Asserts that the given transaction is not null, otherwise prints a
+  /// stack trace to stderr.
+  static void assertHasTransaction(const cling::Transaction* transaction) {
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/DeclCollector.cpp]: Using just T as a variable name is a common notion for variable names of type cling::Transaction.

File: interpreter/cling/lib/Interpreter/DeclCollector.cpp [hunk 2]
```diff
@@ -57,6 +58,15 @@ namespace {
...
+    if (transaction == nullptr) {
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/DeclCollector.cpp]: A more common pattern would be if (!T)

File: interpreter/cling/lib/Interpreter/DeclCollector.cpp [hunk 3]
```diff
@@ -57,6 +58,15 @@ namespace {
...
+      llvm::sys::PrintStackTrace(llvm::errs());
+      assert(false && "Missing transaction during deserialization!");
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/DeclCollector.cpp]: Shouldn't we use llvm_unreachable here. If so we will need to rename the routine.

Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/DeclCollector.cpp]: The original code was using `assert`. For me this looks like a poster child of `assert`: validating a precondition is met.

Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DeclCollector.cpp]: According to some [rules](http://llvm.org/docs/CodingStandards.html#assert-liberally) `llvm_unreachable` is the preferred primitive. I do not have a strong opinion here, but the closer to the llvm coding guidelines we are in cling, the better.

Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/DeclCollector.cpp]: Good points, Vassil. Thanks for explaining!

**Comment by Teemperor** [comment]
Added do not merge as Jenkins is restating so this only looks green.

**Comment by Teemperor** [comment]
@phsft-bot build

**Comment by Teemperor** [comment]
@phsft-bot build

**Comment by Teemperor** [comment]
@phsft-bot build

**Pull request merged**